basis/http/client/client.factor

```diff
@@ -111,15 +142,73 @@ SYMBOL: redirects
     "transfer-encoding" header "chunked" =
     [ read-chunked ] [ each-block ] if ; inline
 
+: request-socket-endpoints ( request -- physical logical )
+    [ proxy-url>> ] [ url>> ] bi [ or ] keep ;
+
 : <request-socket> ( -- stream )
-    request get url>> url-addr ascii <client> drop
+    request get request-socket-endpoints [ url-addr ] bi@
+    remote-address set ascii <client> local-address set
     1 minutes over set-timeout ;
 
+: https-tunnel? ( request -- ? )
+    [ proxy-url>> ] [ url>> protocol>> "https" = ] bi and ;
```

Comment:
I forgot to mention that check-proxy sets proxy-url to f if the url doesn't have a host, so in ```https-tunnel?``` it's never an empty url: either f, or a url with a host.

- If there is no port but there is a protocol:
  - for "http://localhost" (ie the http protocol and no port), then port 80 is used (instead of the more standard 8080 or 3128, I'm not sure what do to in this case..).
  - On linux, for T{ url { protocol "http-alt" } { host "localhost" } }, port 8080 is used
    - Note: url parsing doesn't work on "http-alt://localhost/" >url or "http-alt://localhost" >url, they interpret it as the whole path..
  - Note: On linux, there is no entry for port 3128 in /etc/services, so you can't do something like  T{ url { protocol "squid" } { host "localhost" } }
- If there is no protocol but there is a port:
  - if the user set a string (for example by using the environnement variables where you can't use factor's TUPLE litteral syntax), "localhost:3128" >url parses as T{ url { path "localhost:3128" } } so there is no host and it doesnt use the proxy (so that's not very user friendly..but because of >url?). In addition to checking for a host in proxy-url, maybe we should ensure that the path is empty ?
  - if the user set a url, like URL" //localhost:3128/" or  T{ url { host "localhost" } { port 3128 } }, it "works". That's because we always do HTTP proxy, we don't look at the protocol from the proxy-url (except for guessing a port if missing). If we plan to support more proxy types (ie SOCKS proxy), then we can start doing something based on the protocol of proxy-url..
- if there is no protocol and no port, it throws an obscure exception deep inside the call stack.  ?

Okay so that was long... In the end, what matters is what errors should we send as a friendly message, what errors should we throw as the resulting exception (or even surprising behavior)